src/opm/utility/EModel.cpp

```diff
@@ -0,0 +1,474 @@
+    if ( Opm::filesystem::exists(rootN + ".EGRID") ){
+        hasEgrid = true;
+        gridfile = std::make_unique<Opm::EclIO::EGrid>(rootN+".EGRID");
+    }
+
+    initfile = std::make_unique<Opm::EclIO::EclFile>(rootN+".INIT");
+
+    if (!initfile->hasKey("INTEHEAD")) {
+        std::string message="parameter INTEHEAD not found in selected init file " + rootN+".INIT";
+        throw std::invalid_argument(message);
+    }
+
+    std::vector<int> inteh=initfile->get<int>("INTEHEAD");
+
+    nI = inteh[8];
+    nJ = inteh[9];
+    nK = inteh[10];
+
+    nActive = static_cast<unsigned int>(inteh[11]);
+
+    if (!initfile->hasKey("PORV")) {
+        std::string message="parameter PORV not found in selected init file " + rootN+".INIT";
+        throw std::invalid_argument(message);
+    }
+
+    PORV.reserve(nActive);
+
+    I.reserve(nActive);
+    J.reserve(nActive);
+    K.reserve(nActive);
+
+    ActFilter.resize(nActive, true);
+
+    std::vector<float> porv_all = initfile->get<float>("PORV");
+
+    int n = 0;
+
+    for (int k = 0; k < nK; k++)
+        for (int j = 0; j < nJ; j++)
+            for (int i = 0; i < nI; i++) {
+                if (porv_all[n] > 0.0) {
+                    PORV.push_back(porv_all[n]);
+                    I.push_back(i+1);
+                    J.push_back(j+1);
+                    K.push_back(k+1);
+                }
+                n++;
+            }
+
+    int index = 0;
+
+    initParam["PORV"] = index;
+    initParamName.push_back("PORV");
+    initParamType.push_back(Opm::EclIO::REAL);
+    indInInitEclfile.push_back(-1);
+    index++;
+
+    if (hasEgrid == true){
+        initParam["CELLVOL"] = index;
+        initParamName.push_back("CELLVOL");
+        initParamType.push_back(Opm::EclIO::REAL);
+        indInInitEclfile.push_back(-1);
+        index++;
+    }
+
+    auto initArrList = initfile->getList();
+
+    for (size_t m = 0; m < initArrList.size(); m++) {
+        std::string name = std::get<0>(initArrList[m]);
+        auto arrType = std::get<1>(initArrList[m]);
+        auto sizeArray=std::get<2>(initArrList[m]);
+
+        if (static_cast<size_t>(sizeArray) == nActive) {
+            initParam[name] = index;
+            initParamName.push_back(name);
+            initParamType.push_back(arrType);
+            indInInitEclfile.push_back(m);
+            index++;
+        }
+    }
+
+    if ( Opm::filesystem::exists( rootN + ".UNRST" ) )
+    {
+        rstfile = std::make_unique<Opm::EclIO::ERst>(rootN + ".UNRST");
+        std::vector<int> rstepList = rstfile->listOfReportStepNumbers();
+
+        if (rstepList.size() == 0)
+            throw std::runtime_error("selected restart file have no report steps");
+
+        activeReportStep = rstepList[0];
+        initSolutionData(activeReportStep);
+
+    } else {
+        activeReportStep = -1;
+    }
+
+    celVolCalculated = false;
+    activeFilter = false;
+}
+
+void EModel::setReportStep(int rstep)
+{
+    if (activeReportStep == -1)
```

Comment:
Try to avoid use of magic-invalid-values - just check the rstfile pointer/optional:
```
if (!rstfile.has_value())
    throw ...
```